src/System.IO.Pipelines/src/System/IO/Pipelines/StreamPipeReader.cs

```diff
@@ -362,5 +408,127 @@ private void Cancel()
         {
             InternalTokenSource.Cancel();
         }
+
+        static void ThrowConcurrentReadsNotSupported()
+        {
+            throw new InvalidOperationException($"Concurrent reads are not supported; await the {nameof(ValueTask<ReadResult>)} before starting next read.");
+        }
+
+        private class AsyncReader : IValueTaskSource<ReadResult>
+        {
+            private readonly StreamPipeReader _reader;
+            private readonly Action _onCompletedAction;
+
+            private ManualResetValueTaskSourceCore<ReadResult> _mrvts;
+
+            private ValueTaskAwaiter<int> _readAwaiter;
+            private CancellationTokenSource _tokenSource;
+            private CancellationToken _cancellationToken;
+
+            public AsyncReader(StreamPipeReader streamPipeReader)
+            {
+                _reader = streamPipeReader;
+                _onCompletedAction = new Action(OnCompleted);
+            }
+
+            public ValueTask<ReadResult> CompleteReadAsync(ValueTask<int> task, CancellationTokenSource tokenSource, CancellationToken cancellationToken)
+            {
+                Debug.Assert(_reader._readInProgress, "Read not in progress");
+
+                _tokenSource = tokenSource;
+                _cancellationToken = cancellationToken;
+
+                _readAwaiter = task.GetAwaiter();
+
+                return new ValueTask<ReadResult>(this, _mrvts.Version);
+            }
+
+            private void OnCompleted()
+            {
+                try
+                {
+                    int length = _readAwaiter.GetResult();
+                    _cancellationToken.ThrowIfCancellationRequested();
```

Comment:
This isn’t required. Get result will throw appropriately

Reply:
Fixed